### ISSUE TITLE
Enhancement: Turn rule-schema-to-typescript-types into a public reusable package

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10469]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Coming over from https://github.com/eslint/eslint/issues/18912: @fasttime is trying out `@typescript-eslint/rule-schema-to-typescript-types` on core ESLint rules. Which I think is great - the package has theoretical use outside just typescript-eslint!

The package is marked as private now and so is not published to npm. But I think we should make it public so consumers such as ESLint core can use it.

### Additional Info

Per https://github.com/eslint/eslint/issues/18912#issuecomment-2525114385, the package does have a few limitations at the moment:

* It doesn't include comments, which IMO could be useful
* Some core rules, e.g. `logical-assignment-operators`, use schemas in a way it doesn't support

If we do make it public, those and other limitations would be good followup issues IMO.

Another thought could be: should we split it out of the typescript-eslint monorepo into its own dedicated repository? IMO that would also be a followup discussion. Having it in this repo is nice. It lets us share infra with the rest of the monorepo's packages, and keep its development & package releases in sync with them.

:sparkling

**Comment by bradzacher** [comment]
My plan was always to productionise the tool and the Infra around it.

The goal was to create a standard where all plugins used the tool to generate type decls for the plugin with fully correct types for each rule. Then our config function could use those types to allow TS to check users rule options.

So I'm in favour of making it public.
But I think if we're gonna do that we would want to properly productionise it so that a plugin can just go like "here's my repo generate a `plugin.d.ts` for me."

We needn't design the infra to do config typechecking yet - but generating a `plugin.d.ts` I think is the minimum the package should support for people. I.e. Provide a CLI which accepts a path to either an `index.js` or a `index.ts` etc and then can get the plugin from that, iterate the rules, and generate a type for the plugin.

**Comment by bradzacher** [comment]
Side note: why not use [`json-schema-to-typescript`](https://github.com/bcherny/json-schema-to-typescript)? Because I found it didn't generate good types for our usecase. And there were a number of things I wanted to change about its output. And I also struggled to work with their codebase as it was designed to generically support all json schemas.

Instead our util was intentionally very constrained and purposely did not support all things representable by json schemas because that is a door to insanity. Instead it supports what our rules used.

If there are things in the core repo that it doesn't support - we can build it. Or perhaps those schemas can be changed to a better structure that the tool can support? Would be worth taking a look to decide which way to go. I'd say that if the tool doesn't currently support it then it might generate a dog's breakfast of types...

**Comment by fasttime** [comment]
Thanks for opening this discussion! I was indeed planning to use `rule-schema-to-typescript-types` to autogenerate types for ESLint core rules when I have a chance. I'm not sure if `rule-schema-to-typescript-types` will be able to handle all the specific requirements that ESLint will (probably) run into, which is why I didn't rule out the possibility of using a fork. But before forking `rule-schema-to-typescript-types` to customize one or two things, I would like to contribute to this project as much as possible.

Here is a rough list of TODOs that I think will be necessary before we can start using this in ESLint. I'd be happy to work on all of these changes if accepted:

1. Add unit tests.
2. Fix bugs that appear while testing with ESLint rules. This will include fixing the syntax tree or formatting for schemas that aren't currently translated correctly or throwing an error for schemas that aren't supported yet. The latter should be always the case — which is great, because it means that we can trust the result is correct if one is produced — but maybe something slipped through.
2. Provide an API to modify the generated syntax tree before it is formatted. This is arguably not strictly necessary but it could come in handy with some extra wishes that aren't currently covered by metadata:
   - Add descriptions for options (if desired).
   - Mark options as deprecated.
   - Indicate default values for options that for technical reasons aren't encoded in `meta.defaultOptions`.
1. Publish the package.

Further steps could be providing a command-line interface, adding support for schemas that aren't already covered by the present functionality, or even extending support to all properties of a plugin object as proposed by @bradzacher. Ideally, when all that is done, the process of generating types for ESLint rules will be completely automatic, and we'll get a `rules.d.ts` file with a header like `// Do not manually change. All changes will be overwritten.`

**Comment by JoshuaKGoldberg** [comment]
Coming back to this: I think this is a great area of work and very tricky. Sorry for the delay, I think we'd wanted to prioritize getting to this but haven't had the budget this year. @fasttime all the points you mentioned seem very reasonable - accepting PRs 🙂
